[SPARK-34133][AVRO] Respect case sensitivity when performing Catalyst-to-Avro field matching

xkrogen opened the pull request:
### What changes were proposed in this pull request?
Make the field name matching between Avro and Catalyst schemas, on both the reader and writer paths, respect the global SQL settings for case sensitivity (i.e. case-insensitive by default). `AvroSerializer` and `AvroDeserializer` share a common utility in `AvroUtils` to search for an Avro field to match a given Catalyst field.

### Why are the changes needed?
Spark SQL is normally case-insensitive (by default), but currently when `AvroSerializer` and `AvroDeserializer` perform matching between Catalyst schemas and Avro schemas, the matching is done in a case-sensitive manner. So for example the following will fail:
```scala
      val avroSchema =
        """
          |{
          |  "type" : "record",
          |  "name" : "test_schema",
          |  "fields" : [
          |    {"name": "foo", "type": "int"},
          |    {"name": "BAR", "type": "int"}
          |  ]
          |}
      """.stripMargin
      val df = Seq((1, 3), (2, 4)).toDF("FOO", "bar")

      df.write.option("avroSchema", avroSchema).format("avro").save(savePath)
```

The same is true on the read path, if we assume `testAvro` has been written using the schema above, the below will fail to match the fields:
```scala
df.read.schema(new StructType().add("FOO", IntegerType).add("bar", IntegerType))
  .format("avro").load(testAvro)
```

### Does this PR introduce _any_ user-facing change?
When reading Avro data, or writing Avro data using the `avroSchema` option, field matching will be performed with case sensitivity respecting the global SQL settings.

### How was this patch tested?
New tests added to `AvroSuite` to validate the case sensitivity logic in an end-to-end manner through the SQL engine.

xkrogen committed:
Respect case-insensitivity of schema matching when writing Avro with a user-provided schema. Enhance error messages on schema mismatch.

xkrogen committed:
Merge AvroSerializerSuite and AvroDeserializeSuite into AvroSerdeSuite since they were testing very similar things.

xkrogen committed:
Enhance AvroDeserializer to separately track avroPath and catalystPath since they may differ. Standardize exception messages between serialize and deserialize paths. Minor fixups.

SparkQA commented:
**[Test build #134128 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134128/testReport)** for PR 31201 at commit [`ab71db2`](https://github.com/apache/spark/commit/ab71db27a8a38cf27c8bf35dc41630ee9cc8eb7e).

SparkQA commented:
**[Test build #134128 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134128/testReport)** for PR 31201 at commit [`ab71db2`](https://github.com/apache/spark/commit/ab71db27a8a38cf27c8bf35dc41630ee9cc8eb7e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

xkrogen committed:
Fix Scala style issues and resolve compilation issues on Scala 2.13

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38711/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134128/

xkrogen commented:
cc @dbtsai @viirya @gengliangwang @cloud-fan

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38711/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38711/

SparkQA commented:
**[Test build #134131 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134131/testReport)** for PR 31201 at commit [`8ae6b7f`](https://github.com/apache/spark/commit/8ae6b7fba447028e715347e919636c243b8b52fd).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38714/

SparkQA commented:
**[Test build #134131 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134131/testReport)** for PR 31201 at commit [`8ae6b7f`](https://github.com/apache/spark/commit/8ae6b7fba447028e715347e919636c243b8b52fd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134131/

cloud-fan commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:
```diff
@@ -330,27 +341,27 @@ private[sql] class AvroDeserializer(
     var i = 0
     while (i < length) {
       val sqlField = sqlType.fields(i)
-      val avroField = avroType.getField(sqlField.name)
```
is the original code `avroType.getField(sqlField.name)` also a O(n) linear lookup?

xkrogen replied:
No, it is a constant-time map lookup (from `org.apache.avro.Schema.RecordSchema`):
```
    public Field getField(String fieldname) {
      if (fieldMap == null)
        throw new AvroRuntimeException("Schema fields not set yet");
      return fieldMap.get(fieldname);
    }
```
I think this is not so performance-critical given that it is executed only once per query? It is not done on a per-record basis.

xkrogen committed:
Test fixups for AvroSuite to accommodate new error messages

xkrogen committed:
Minor fixup to AvroSuite.'Ignore corrupt Avro file if flag IGNORE_CORRUPT_FILES enabled' to make it succeed in a broader variety of environments.

SparkQA commented:
**[Test build #134233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134233/testReport)** for PR 31201 at commit [`28da8da`](https://github.com/apache/spark/commit/28da8da9115fa0acdd64a4991ef96b13f518eb01).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38818/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38818/

SparkQA commented:
**[Test build #134233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134233/testReport)** for PR 31201 at commit [`28da8da`](https://github.com/apache/spark/commit/28da8da9115fa0acdd64a4991ef96b13f518eb01).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134233/

cloud-fan commented:
Can we improve the error message in a separated PR? This PR is a bit hard to review as many diffs are not related to the case sensitivity fix.

xkrogen commented:
Yes, sure. The error message changes did grow fairly large in the end. I will post a separated PR later today.

xkrogen committed:
Revert error message-related changes, leaving only case sensitivity changes.

xkrogen commented:
Thanks for the suggestion @cloud-fan, I have updated the PR branch, updated the description of the PR/JIRA, and created a new JIRA for the error messages: SPARK-34182

SparkQA commented:
**[Test build #134293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134293/testReport)** for PR 31201 at commit [`6ca9247`](https://github.com/apache/spark/commit/6ca92479138eef58b46d67818a52fab6e6c91dec).

SparkQA commented:
**[Test build #134293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134293/testReport)** for PR 31201 at commit [`6ca9247`](https://github.com/apache/spark/commit/6ca92479138eef58b46d67818a52fab6e6c91dec).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134293/

dongjoon-hyun commented:
Retest this please.

dongjoon-hyun commented:
Since the master branch upgrades Avro to 1.10.1, I re-trigger the tests.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38879/

SparkQA commented:
**[Test build #134295 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134295/testReport)** for PR 31201 at commit [`6ca9247`](https://github.com/apache/spark/commit/6ca92479138eef58b46d67818a52fab6e6c91dec).

SparkQA commented:
**[Test build #134295 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134295/testReport)** for PR 31201 at commit [`6ca9247`](https://github.com/apache/spark/commit/6ca92479138eef58b46d67818a52fab6e6c91dec).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38881/

cloud-fan commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:
```diff
@@ -201,4 +201,25 @@ private[sql] object AvroUtils extends Logging {
+   * @param catalystField The catalyst field for which to search for a match.
+   * @return `Some(match)` if a matching Avro field is found, otherwise `None`.
+   */
+  private[avro] def getAvroFieldForCatalystField(
```
This has nothing to do with catalyst field, only the name, how about
```
def getAvroFieldByName(avroSchema: Schema, name: String): Option[Schema.Field] = {
  val isEqual = SQLConf.get.resolver
  avroSchema.getFields.asScala.filter ...
}
```

gengliangwang replied:
+1

xkrogen replied:
Completely right, good call. Also thanks for the pointer on `SQLConf.resolver`, that is much nicer!

cloud-fan commented on external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:
```diff
@@ -1261,6 +1262,58 @@ abstract class AvroSuite
     }
   }
 
+  test("Reading with user provided schema respects case sensitivity for field matching") {
```
let's put the jira id in the test name: `test("SPARK-xxxx: ....")`

cloud-fan commented on external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:
```diff
@@ -1261,6 +1262,58 @@ abstract class AvroSuite
+    }
+  }
+
+  test("Writing with user provided schema respects case sensitivity for field matching") {
```
ditto

gengliangwang commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:
```diff
@@ -201,4 +201,25 @@ private[sql] object AvroUtils extends Logging {
+      case Seq(avroField) => Some(avroField)
+      case Seq() => None
+      case s => throw new IncompatibleSchemaException(
+        s"Searching for ${catalystField.name} in Avro schema gave ${s.size} matches")
```
Nit: Let's print the Avro field names in the error message.

gengliangwang replied:
Also, we need a test case for this

xkrogen replied:
Good idea on printing Avro field names! I added a test case as well.

xkrogen committed:
Change AvroUtils new function to getAvroFieldByName to be more generic. Add a test for case-insensitive matching multiple Avro fields.

SparkQA commented:
**[Test build #134341 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134341/testReport)** for PR 31201 at commit [`25b0a39`](https://github.com/apache/spark/commit/25b0a3943b960081ec7647e54a80155e3905f78e).

SparkQA commented:
**[Test build #134341 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134341/testReport)** for PR 31201 at commit [`25b0a39`](https://github.com/apache/spark/commit/25b0a3943b960081ec7647e54a80155e3905f78e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38928/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38928/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38928/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134341/

gengliangwang commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:
```diff
@@ -201,4 +203,32 @@ private[sql] object AvroUtils extends Logging {
+   *                                     is used and `avroSchema` has two or more fields that have
+   *                                     the same name with difference case).
+   */
+  private[avro] def getAvroFieldByName(avroSchema: Schema, name: String):
```
As per https://github.com/databricks/scala-style-guide, let's make it:
```
private[avro] def getAvroFieldByName(
    avroSchema: Schema,
    name: String): Option[Schema.Field] = {
```

xkrogen replied:
I think the way I have it now is valid under the style guide:

> For method declarations, use 4 space indentation for their parameters and put each in each line when the parameters don't fit in two lines. Return types can be either on the same line as the last parameter, or start a new line with 2 space indent.

Example:
```
def newAPIHadoopFile[K, V, F <: NewInputFormat[K, V]](
    path: String,
    fClass: Class[F],
    kClass: Class[K],
    vClass: Class[V],
    conf: Configuration = hadoopConfiguration)
  : RDD[(K, V)] = {
  // method body
}
```

It does seem that the colon is supposed to go on the second line, though. I need to update it anyway and don't have much of an opinion about which is cleaner so I'll go with your proposal.

gengliangwang reviewed (approved):
LGTM except one comment

xkrogen committed:
Minor scala style fix

xkrogen commented:
Thanks for the reviews @gengliangwang ! Should be all good to go now.

SparkQA commented:
**[Test build #134377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134377/testReport)** for PR 31201 at commit [`d919613`](https://github.com/apache/spark/commit/d919613197115bf08a8b0be15577df3a595d66b7).

SparkQA commented:
**[Test build #134377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134377/testReport)** for PR 31201 at commit [`d919613`](https://github.com/apache/spark/commit/d919613197115bf08a8b0be15577df3a595d66b7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38964/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134377/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38964/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38964/

cloud-fan commented:
thanks, merging to master/3.1!

xkrogen commented:
Many thanks @cloud-fan !
And thank you @gengliangwang for the reviews!

dongjoon-hyun commented:
+1, late LGTM. Thank you all!
